File: jmlee/week5/maximumsubarray/maximumsubarray.go
package maximumsubarray

import "fmt"

/*
Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

Example:

Input: [-2,1,-3,4,-1,2,1,-5,4],
Output: 6
Explanation: [4,-1,2,1] has the largest sum = 6.
Follow up:

If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
*/

func maxSubArray(nums []int) int {
	len := len(nums)

	var maxSum int
	var maxSubArray []int

	for i := 1; i <= len; i++ {
		chunkSize := i
		for start := 0; start <= len-chunkSize; start++ {
			subArray := nums[start : start+chunkSize]
			subArraySum := sum(subArray)
			if subArraySum > maxSum || maxSubArray == nil {
				maxSum = subArraySum
				maxSubArray = subArray
			}
		}
	}

	fmt.Printf("%v -> sum : %v\n", maxSubArray, maxSum)
	return maxSum
}

func sum(nums []int) int {
	sum := 0
	for _, value := range nums {
		sum += value
	}

	return sum
}

File: carrotfarm/src/p2018_08_14/MaximumSubarray.java
package p2018_08_14;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

//If you have figured out the O(n) solution,
// try coding another solution using the divide and conquer approach, which is more subtle.

public class MaximumSubarray {

	public int maxSubArray(int[] nums) {

		int totMax = nums[0];
		int tmpMax = nums[0];

		for (int i=1; i< nums.length; i++) {

			int val = nums[i];
			if (val > tmpMax || val > 0) {
				tmpMax = Math.max(val, tmpMax+val);
			}else {
				tmpMax = tmpMax + val;
			}
			if (tmpMax > totMax) {
				totMax = tmpMax;
			}
		}
		return  totMax;
	}

	@Test
	public void test() {
//		4,-1,2,1
		assertEquals(maxSubArray(new int[] {-2,1,-3,4,-1,2,1,-5,4}), 6);
	}
}
